blocks/graph.py

```diff
@@ -37,6 +37,7 @@ def _get_variables(self):
         self.applies = set()
         self.application_calls = set()
         self.updates = []
+        self.auxiliary_variables = set()
```

Comment:
Although the interface looks okay, my worry is about getting the correct the parameters out, because this means knowing for each brick which of its parameters are weights. Do we want to start tagging parameters with their roles e.g. `VariableRole.WEIGHTS` and `VariableRole.BIASES`? Otherwise how is a user expected to e.g. select the 3rd and 7th set of weights in a 20-layer deep autoencoder created through the `MLP` class.

Although I appreciate the argument about splitting things up into parts, I also don't like the idea of some regularizers being done this way, and others in yet another way. It would be nice to have a single unified interface for that, especially since the interface is pretty complex already; it will never be as easy as Pylearn2's `model.get_weight_decay()`. The great thing of attaching regularizers and other costs to application calls then, is that you can do some crazy things. Let's say it's now a thing to penalize the L2 difference between successive weight matrices in a deep autoencoder, but you want to exclude the middle layers from this. For an MLP, this would be easy by using the application call:
```python
for num, (i, j) in enumerate(zip(self.linear_transformations[:-1],
﻿                                 self.linear_transformations[1:])):
    application_call.add_auxiliary_variable(
        (i.W - j.W).norm(2), roles=[VariableRole.COST],
        name='L2_difference_{}'.format(num))
```
```python
costs = cg.get_variables(roles=[VariableRole.COST])
costs = [cost for cost in costs if cost.name != 'L2_difference_9']
```
The alternative would be to write a new function, `add_successive_L2_regularization(param)`*, but now we're stuck with the problem of getting the parameters out again.

I think it would be nice if we could come up with a very specific example of what a complicated case would look like exactly. Let me give this as a first attempt (attempts, really):

```python
mlp = MLP(activations=[Tanh() for _ in range(4)], dimensions=[np.random.randint(11)])
y_hat = mlp.apply(x)
cost = CategoricalCrossEntropy().apply(y, y_hat)

cg = ComputationGraph([y_hat])
regularizers = cg.get_variables(roles=[VariableRole.COST])
# Maybe it would be nicer if start using roles more heavily,
# so that we could say e.g. cg.get_variables(roles=[VariableRole.L2_WEIGHT_DECAY])
l2_weight_decay = [var for var in regularizers if 'L2' in var.name]
# This would require variables to be returned in topological order
l2_weight_decay = l2_weight_decay[:5] + l2_weight_decay[6:]
# Probably cleaner
l2_weight_decay = [var for var in l2_weight_decay
                   if var.tag.brick is not mlp.linear_transformations[6]]
```
I guess that this is which is where I started wondering whether maybe we should stop using selectors on bricks and select on the computation graph itself instead. This is in line with out idea of being more of a Theano framework, and would make best use of the annotation. Of course, we could have both: Selectors for bricks, and for the computation graph. Then we could do something like.
```python
# This or a more fancy kind of matching language like Jan's if you will
weight_norms = cg.get_variables(roles=[VariableRole.L2_WEIGHT_DECAY],
                                exclude_bricks=[mlp.linear_transformations[5]])
```

*Not sure why you wanted to return the cost directly, it seems easier to return just the regularization term so that the user can add that to the cost himself, potentially with a variable regularization coefficient.